### PR TITLE
linalg: QR factorization

### PR DESCRIPTION
**Pull request by perazz**
Compute the QR factorization of a `real` or `complex` matrix: $A = Q R $, where q is orthonormal and r is upper-triangular. Matrix $A$ has size `[m,n]`, with $m\ge n$. 
Based on LAPACK General QR factorization (`*GEQRF`) and ordered matrix output (`*ORGQR`, `*UNGQR`). 
Option for `full` or `reduced` factorization: given `k = min(m,n)`, one can write $A = \( Q_1  Q_2 \) \cdot \( \frac{R_1}{0}\)$. The user may want the full problem or the reduced problem only $A = Q_1 R_1 $.

- [x] base implementation
- [x] tests
- [x] ~~exclude unsupported `xdp`~~
- [x] documentation  
- [x] submodule 
- [x] examples
- [x] all `pure subroutine` interfaces
- [x] workspace size evaluation 

### Prior art
- Numpy: `linalg.qr(a, mode={'reduced', 'complete', 'r', 'raw')`
- Scipy: `scipy.linalg.qr(a, overwrite_a=False, lwork=None, mode='full', pivoting=False, check_finite=True)`

### Proposed implementation

- `call qr(A,Q,R [, overwrite_a] [, storage] [, err])` : `pure` subroutine interface
- `call qr_space(A, lwork [, err])` query internal storage size for pre-allocation. 

1) Special care was devoted to re-using internal storage such that a `pure` and totally allocation-less method is available, if the user provides pre-allocated working array `storage`. Because LAPACK internals require two steps ("raw" factorization + ordered matrix output), temporary storage is borrowed from either `a`, `q`, or `r` during the operations. 

2) Both NumPy and SciPy have a cryptic UI that requires to provide a "method": `full`, `reduced` `raw` `economic`, `r`. I believe this is hard to understand. Instead, the current version proposes to decide it autonomously: 

- If `shape(Q)==[m,m]` and `shape(R)==[m,n]`, perform a full factorization, $A=QR$.
- If `shape(Q)==[m,k]` and `shape(R)==[k,n]`, perform a "reduced" factorization, $A=Q_1R_1$.
- On insufficient/unfit size(s), raise an error.

In other words, the user decides what method is required based on the size of the arrays passed to `qr`.

cc: @fortran-lang/stdlib @jvdp1 @jalvesz @everythingfunctional

### CODE REVIEW

**Comment by perazz** [comment]
From Fortran Montly call: 
- [ ] require that `Q` and `R` have exact sizes for either the full or the reduced problem (do not just check that there is "enough" storage)

**Review by jvdp1 — COMMENTED** [review]
Here are a few comments. LGTM. Thank you

File: src/stdlib_linalg.fypp [hunk 1]
```diff
@@ -442,6 +444,75 @@ module stdlib_linalg
...
+         !> Input matrix a[m,n]
+         ${rt}$, intent(inout), target :: a(:,:)
+         !> Orthogonal matrix Q ([m,m], or [m,k] if reduced)
+         ${rt}$, intent(out), contiguous, target :: q(:,:)
+         !> Upper triangular matrix R ([m,n], or [k,n] if reduced)
+         ${rt}$, intent(out), contiguous, target :: r(:,:)
+         !> [optional] Can A data be overwritten and destroyed?
+         logical(lk), optional, intent(in) :: overwrite_a
+         !> [optional] Provide pre-allocated workspace, size to be checked with qr_space
+         ${rt}$, intent(inout), optional, target :: storage(:)
```
[the remark this answers is not in the document]
Comment (perazz) [reply src/stdlib_linalg.fypp:484]: The storage is used internally as a temporary by LAPACK, and must be `intent(inout)`. But you're right about the next occurrence: the matrix can be just `intent(in)`.

[the remark this answers is not in the document]
Comment (jvdp1) [reply src/stdlib_linalg.fypp:484]: Sorry, I meant `intent(out)` for `storage` (as its content is not used). By I am fine with `intent(inout)`.

[the remark this answers is not in the document]
Comment (perazz) [reply src/stdlib_linalg.fypp]: Oh - yes, absolutely! Here it is: [8827bf7](https://github.com/fortran-lang/stdlib/pull/832/commits/8827bf77f61f838656ed2ad1570a5e0f6043a063)

File: src/stdlib_linalg_qr.fypp [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+         end select
+
+     end subroutine handle_geqrf_info
+
+     #:for rk,rt,ri in RC_KINDS_TYPES
+     
+     ! Get workspace size for QR operations
+     pure module subroutine get_qr_${ri}$_workspace(a,lwork,err)
+         !> Input matrix a[m,n]
+         ${rt}$, intent(inout), target :: a(:,:)
```
[the remark this answers is not in the document]
Comment (perazz) [reply src/stdlib_linalg_qr.fypp]: Yes, `*getrf` wants `intent(inout)` but we can use a dummy variable as the matrix is not accessed when querying the storage space. Here it is: [e69e767](https://github.com/fortran-lang/stdlib/pull/832/commits/e69e767e9941045b15473e2129595754265f7370)

File: doc/specs/stdlib_linalg.md [hunk 3]
```diff
@@ -901,6 +901,82 @@ Exceptions trigger an `error stop`.
...
+Experimental
+
+### Description
+
+This subroutine computes the QR factorization of a `real` or `complex` matrix: \( A = Q R \). where \( Q \) 
+is orthonormal and \( R \) is upper-triangular. Matrix \( A \) has size `[m,n]`, with \( m \ge n \). 
+
+The results are returned in output matrices \( Q \) and \(R \), that have the same type and kind as \( A \). 
+Given `k = min(m,n)`, one can write \( A = \( Q_1  Q_2 \) \cdot \( \frac{R_1}{0}\) \). 
+Because the lower rows of \( R \) are zeros, the user may require to return full matrices (provide `shape(Q)==[m,m]`, `shape(R)==[m,n]`)
```
[the remark this answers is not in the document]
Comment (perazz) [reply doc/specs/stdlib_linalg.md]: Thanks. Improved here: [e3db36d](https://github.com/fortran-lang/stdlib/pull/832/commits/e3db36ddc5067f6f971d04e9b93644eed8146c10)

File: test/linalg/test_linalg_qr.fypp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        #:else
+        a = rea
+        #:endif
+        aorig = a
+        
+        ! 1) QR factorization with full matrices. Input NaNs to be sure Q and R are OK on return
+        q = 0.0_${rk}$
+        q = 1.0_${rk}$/q
+        r = 0.0_${rk}$
+        r = 1.0_${rk}$/r
```
Comment (jalvesz) [inline test/linalg/test_linalg_qr.fypp:57]: Why not use the ieee module?
```suggestion
        q = ieee_value(0.0_${rk}$,  ieee_quiet_nan)
        r = ieee_value(0.0_${rk}$,  ieee_quiet_nan)     
```

**Review by jalvesz — CHANGES_REQUESTED** [review]

File: test/linalg/test_linalg_qr.fypp (hunk 4 above)
Comment (perazz) [reply test/linalg/test_linalg_qr.fypp]: [6fc0e1d](https://github.com/fortran-lang/stdlib/pull/832/commits/6fc0e1d87465723abb499dbdd58b034379a793a2)

**Comment by perazz** [comment]
Thank you for your edits @jvdp1 @jalvesz, looks much better now.

**Review by jvdp1 — APPROVED** [review]
LGTM. Thank you @perazz for this addition.

**Comment by perazz** [comment]
Thank you @jvdp1. I think it makes sense to wait for a bit longer, and then merge if there are no further comments.

**Comment by perazz** [comment]
Thank you @jvdp1 @jalvesz, I will merge due to no outstanding comments.
